### ISSUE TITLE
RPC state methods unable to handle StorageKey as input

### ISSUE THREAD

**Issue by Stefie** [polkadot-js/api#1239]
I noticed that  the RPC state calls can't handle the input of type `StorageKey` anymore.
I tried with different methods like `getChildKeys`, `getChildStorage`, `getStorage`, `getStorageHash` etc.

Every RPC call that accepts `StorageKey` as a parameter type fails when passing the data as `StorageKey`. The same calls work fine when I'm calling the `toHex()`/ `toString()` methods on the values.

An example to replicate this issue:
```
// get storage keys for ':code' storage
const keys = await rpc.state.getKeys('0x3a636f6465').toPromise(); 

// The first storage key as `StorageKey`, this fails
const typeStorageKey = keys[0]; 
const storageHash = await rpc.state.getStorageHash(typeStorageKey, '0x').toPromise() 
```
Error message:
<img width="1080" alt="Screenshot 2019-07-30 at 19 58 14" src="https://user-images.githubusercontent.com/125398/62153228-66f3b700-b304-11e9-97b2-102edee886c2.png">

```
// The first storage key hex encoded, this succeeds
const typeHex = keys[0].toHex();
const storageHashHex = await rpc.state.getStorageHash(typeHex, '0x').toPromise() 
```
<img width="534" alt="Screenshot 2019-07-30 at 20 05 03" src="https://user-images.githubusercontent.com/125398/62153652-5859cf80-b305-11e9-85f9-987fe94c4e61.png">

I also noticed, that the `meta, method, outputType, section` are undefined, but I'm not sure if that could have something to do with it.
<img width="546" alt="Screenshot 2019-07-30 at 20 02 14" src="https://user-images.githubusercontent.com/125398/62153521-031dbe00-b305-11e9-9496-0afa70871071.png">

It’s the case on both the latest Polkadot and Substrate.

**Comment by jacogr** [comment]
Both of these should be true, something fails -

- https://github.com/polkadot-js/api/blob/master/packages/types/src/primitive/StorageKey.ts#L59
- https://github.com/polkadot-js/api/blob/master/packages/types/src/primitive/Bytes.ts#L41

Start by logging there, on failure, quite possible that we would need to convert all `instanceof Clazz` to `instanceof ClassOf('Clazz')` (not just in these 2 cases, but start logging here and attempt, if false, to use `ClassOf`)

**Comment by Stefie** [comment]
I already tried to debug it before I wrote the ticket. The `decodeStorageKey` method is called twice when I'm passing a `StorageKey`, once with the bytes of the `StorageKey`and then a second time with the value `4` (of type number), which I just can't make any sense of...

**Comment by jacogr** [comment]
Do you have an e2e test that shows this issue?

**Comment by Stefie** [comment]
> Do you have an e2e test that shows this issue?

In my branch, that's how I found this issue. I just commited my WIP tests here:
https://github.com/polkadot-js/api/pull/1167/files#diff-2441a73200dcfac3e29e93eb84b66adcR103

If you remove the `toHex()`the tests are failing. With toHex they are also failing, but only because I'm checking for the wrong values, but the RPC calls are working fine.

**Comment by Stefie** [comment]
I just added 2 tests without the ink contracts / child-storage overhead. They are failing on all versions.

https://github.com/polkadot-js/api/pull/1167/commits/0b30eee2a1ed64cf967103db814ef2777c4dc238

<img width="686" alt="Screenshot 2019-07-30 at 21 46 25" src="https://user-images.githubusercontent.com/125398/62160158-9c53d100-b313-11e9-9fa1-37c711dc1899.png">

**Comment by jacogr** [comment]
@amaurymartiny So I'm loggin inside the bytes constructor for the StorageKey, trying this -

```js
console.error(
  'decodeBytes',
  value,
  value instanceof U8a,
  value instanceof ClassOf('StorageData'),
  value instanceof ClassOf('StorageKey')
);
```

And it returns 

```    
decodeBytes StorageKey [Uint8Array] [
      58,
      99,
      111,
      100,
      101,
      _meta: undefined,
      _method: undefined,
      _outputType: undefined,
      _section: undefined
    ] true true true
```

Check the true, true, true at the end. Something is wonky here.

Definitions -

- `StorageData: Bytes`
- `export default class StorageKey extends Bytes {`

So can understand StorageKey instanceof U8a (which is the base for Bytes), but not the StorageData link here.

**Comment by jacogr** [comment]
@Stefie This _should_ help... https://github.com/polkadot-js/api/pull/1240 - would appreciate a check

Or at the very least, should be done for the issue I mentioned above with the instances.

**Comment by Stefie** [comment]
solved by #1240
